build/scripts/build-utils.ps1

```diff
@@ -0,0 +1,151 @@
+    if (${env:VSINSTALLDIR} -ne $null) {
+
+        # This line deliberately avoids using -ErrorAction.  Inside a VS command prompt
+        # an MSBuild command should always be available.
+        $p = (get-command msbuild).Source
+        write-output "vscmd"
+        write-output $p
+        return
+    }
+
+    # Look for a valid VS installation
+    try {
+        # It's important to execute locate-vs using a new powershell process.  This script will 
+        # use assemblies from our NuGet cache to locate VS.  This will cause them to be locked on
+        # disk for the duration of the powershell session.  Using a separate process makes sure
+        # they are not locked to this session which can be long lived.
+        $p = & powershell -noprofile -executionPolicy RemoteSigned -file (join-path $PSScriptRoot "locate-vs.ps1")
+        if ($? -and (test-path $p)) {
+            $p = join-path $p "..\..\MSBuild\15.0\Bin"
+            $p = [IO.Path]::GetFullPath($p)
+            write-output "vsinstall"
+            write-output $p
+            return
+        }
+    }
+    catch { 
+        # Failures are expected here when no VS installation is present on the 
+        # machine.
+    }
+
+    write-output "xcopy"
+    write-output (locate-msbuild-xcopy)
+    return
+}
+
+function locate-msbuild() {
+    param ([switch]$xcopy = $false) 
+    $both = locate-msbuild-core -xcopy:$xcopy
+    return $both[1]
+}
+
+# Locate the xcopy version of MSBuild
+function locate-msbuild-xcopy() {
+    # Use the xcopy toolset
+    & (join-path $PSScriptRoot "deploy-msbuild.ps1")
+    $p = join-path $PSScriptRoot "..\..\Binaries\Toolset\msbuild"
+    $p = [IO.Path]::GetFullPath($p)
+    return $p
+}
+
+# Locate the xcopy version of the framework assemblies that are used by MSBuild
+function locate-framework-xcopy() {
+    $msbuildDir = locate-msbuild-xcopy
+    return (join-path $msbuildDir "Framework")
+}
+
+# Return the path to the custom reference assembly folder.  In the case a custom
+# folder is not meant to be used $null will be returned instead.
+function locate-reference-assemblies() {
+    $both = locate-msbuild-core
+    if ($both[0] -eq "xcopy") { 
+        return (locate-framework-xcopy)
+    }
+
+    return $null
+}
+
+# This function will ensure that MSBuild is available on the PATH and that any other
+# environment variables necessary for MSBuild to function are also set.
+function ensure-msbuild() {
+    param ([switch]$xcopy = $false) 
+    $both = locate-msbuild-core $xcopy
+    $msbuildDir = $both[1]
+    switch ($both[0]) {
+        "xcopy" {
+            ${env:PATH}="$msbuildDir;${env:PATH}"
+            ${env:TargetFrameworkRootPath}=(locate-framework-xcopy)
+            break;
+        }
+        "vscmd" {
+            # Nothing to do here as the VS Dev CMD should set all appropriate environment
+            # variables.
+            break;
+        }
+        "vsinstall" {
+            ${env:PATH}="$msbuildDir;${env:PATH}"
```

Comment:
Because launching the developer command prompt does a lot more than just provide MSBuild support, it puts a huge number of other tools / settings into the environment.  That potentially, and in the past concretely, changed the output of the build and tests.  Using MSBuild only, because that's all we actually need, explicitly hepls make the tests more reliable.